xbmc/utils/DisplayInfo.cpp

```diff
@@ -0,0 +1,170 @@
+/*
+ *  Copyright (C) 2023 Team Kodi
+ *  This file is part of Kodi - https://kodi.tv
+ *
+ *  SPDX-License-Identifier: GPL-2.0-or-later
+ *  See LICENSES/README.md for more information.
+ */
+
+#include "DisplayInfo.h"
+
+#include "utils/log.h"
+
+extern "C"
+{
+#include <libdisplay-info/cta.h>
+#include <libdisplay-info/edid.h>
+#include <libdisplay-info/info.h>
+}
+
+using namespace KODI;
+using namespace UTILS;
+
+void CDisplayInfo::DiInfoDeleter::operator()(di_info* p)
+{
+  di_info_destroy(p);
+}
+
+std::unique_ptr<CDisplayInfo> CDisplayInfo::Create(std::vector<uint8_t> edid)
+{
+  struct DisplayInfoMaker : public CDisplayInfo
+  {
+    DisplayInfoMaker(std::vector<uint8_t> edid) : CDisplayInfo(edid) {}
+  };
+
+  auto info = std::make_unique<DisplayInfoMaker>(edid);
+  if (!info->IsValid())
+    return {};
+
+  info->Parse();
+
+  info->LogInfo();
+
+  return info;
+}
+
+CDisplayInfo::CDisplayInfo(std::vector<uint8_t> edid)
+{
+  m_info.reset(di_info_parse_edid(edid.data(), edid.size()));
+}
+
+CDisplayInfo::~CDisplayInfo() = default;
+
+bool CDisplayInfo::IsValid() const
+{
+  if (!m_info)
+    return false;
+
+  const char* error = di_info_get_failure_msg(m_info.get());
+  if (error)
+  {
+    CLog::Log(LOGERROR, "[display-info] error: {}", error);
+    return false;
+  }
+
+  return true;
+}
+
+void CDisplayInfo::Parse()
+{
+  char* str = nullptr;
+
+  str = di_info_get_make(m_info.get());
+  m_make = str ? str : "unknown";
+  free(str);
+
+  str = di_info_get_model(m_info.get());
+  m_model = str ? str : "unknown";
+  free(str);
+
+  const di_edid* edid = di_info_get_edid(m_info.get());
+
+  const di_edid_ext* const* extensions = di_edid_get_extensions(edid);
+  for (const di_edid_ext* extension = *extensions; extension != nullptr;
```

Comment:
I don't love this c-array looping.

Reply:
A (IMHO) more readable way to loop, without using pointer arithmetics:

```c
for (size_t i = 0; extensions[i] != nullptr; i++) {
    const di_edid_ext *extension = extensions[i];
}
```